modules/firebase/firebase.go

```diff
@@ -0,0 +1,165 @@
+func WithData(dataPath string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) error {
+		req.Env["DATA_DIRECTORY"] = dataPath
+		return nil
+	}
+}
+
+const cacheFilePath = "/root/.cache/firebase"
+
+// WithCache enables firebase binary cache based on session (meaningful only when multiple tests are used)
+func WithCache() testcontainers.CustomizeRequestOption {
+	volumeName := fmt.Sprintf("firestore-cache-%s", testcontainers.SessionID())
+	volumeOptions := &mount.VolumeOptions{
+		Labels: testcontainers.GenericLabels(),
+	}
+
+	return func(req *testcontainers.GenericContainerRequest) error {
+		m := testcontainers.ContainerMount{
+			Source: testcontainers.DockerVolumeMountSource{
+				Name:          volumeName,
+				VolumeOptions: volumeOptions,
+			},
+			Target: cacheFilePath,
+		}
+		req.Mounts = append(req.Mounts, m)
+		return nil
+	}
+}
+
+func gatherPorts(config FirebaseConfig) ([]string, error) {
+	var ports []string
+
+	v := reflect.ValueOf(config.Emulators)
+	for i := 0; i < v.NumField(); i++ {
+		emulator := v.Field(i)
+		if emulator.Kind() != reflect.Struct {
+			continue
+		}
+		name := v.Type().Field(i).Name
+
+		enabledF := emulator.FieldByName("Enabled")
+		if enabledF != (reflect.Value{}) && !enabledF.Bool() {
+			continue
+		}
+
+		hostF := emulator.FieldByName("Host")
+		portF := emulator.FieldByName("Port")
+		websocketPortF := emulator.FieldByName("WebsocketPort")
+
+		if hostF != (reflect.Value{}) && !hostF.IsZero() {
+			host := hostF.String()
+			if host != "0.0.0.0" {
+				return nil, fmt.Errorf("config specified %s emulator host on non public ip: %s", name, host)
+			}
+		}
+		if portF != (reflect.Value{}) && !portF.IsZero() {
+			port := fmt.Sprintf("%d/tcp", portF.Int())
+			ports = append(ports, port)
+		}
+		if websocketPortF != (reflect.Value{}) && !websocketPortF.IsZero() {
+			port := fmt.Sprintf("%d/tcp", websocketPortF.Int())
+			ports = append(ports, port)
+		}
+	}
+
+	return ports, nil
+}
+
+// Run creates an instance of the Firebase container type
+func Run(ctx context.Context, img string, opts ...testcontainers.ContainerCustomizer) (*Container, error) {
+	req := testcontainers.GenericContainerRequest{
+		ContainerRequest: testcontainers.ContainerRequest{
+			Image:      img,
+			Env:        map[string]string{},
+			WaitingFor: wait.ForLog("All emulators ready! It is now safe to connect your app."),
+		},
+		Started: true,
+	}
+
+	for _, opt := range opts {
+		if err := opt.Customize(&req); err != nil {
+			return nil, fmt.Errorf("customize: %w", err)
+		}
+	}
+
+	// Check if user supplied root:
+	rootPathIdx := slices.IndexFunc(req.ContainerRequest.Files, func(file testcontainers.ContainerFile) bool {
+		return file.ContainerFilePath == rootFilePath
+	})
+	if rootPathIdx == -1 {
+		return nil, fmt.Errorf("firebase emulator is unable to boot without configuration root (obtained via 'firebase init')")
```

Comment:
suggestion: this seems to indicate there is a missing configuration to the container, could be be more explicit / concise about what the user did wrong?

Reply:
I am a bit conflicted about this error check in general.

Depending on the container, users might have a default root available, so that the emulator would launch even if no external root is mounted/copied.

I guess I am removing it, with the assumption that users can read container boot logs and see what is missing.

Reply:
But, then - the port discovery process breaks... Unless I can inspect the container file system before launch.